third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp

```diff
@@ -542,16 +543,66 @@ SmallVector<unsigned, 2> deduceTilesPerWarpForScale(
   return chosen;
 }
 
+// Detect if transposed layout is needed to enable large vectorization of global
+// stores or intra warp conversion.
+bool needTransposedMfmaLayout(tt::DotOp dotOp, int mfmaVersion, int mDim,
+                              int nDim, bool is16BitElemTy,
+                              bool hasPreShuffledScale,
+                              ArrayRef<unsigned> warpsPerTile,
+                              ttg::AMDMfmaEncodingAttr &mfmaEnc,
+                              AMD::ModuleAxisInfoAnalysis &axisAnalysisPass) {
+  // We always transpose 64x4 mfma in order to use the mfma broadcast.
+  if (mDim == 64 && nDim == 4)
+    return true;
+
+  // We can not support transposed mfma 4x64 as it requires to broadcast the
+  // operand A.
+  if (mDim == 4 && nDim == 64)
+    return false;
+
+  // Set isTransposed based on how the result of dotOp is stored.
+  // If the store’s contiguity aligns with the MFMA layout order, keep the
+  // isTranposed flag set to true. Otherwise, set it to false. This allows us to
+  // vectorize the store along the dimension where elements are accessed
+  // contiguously.
+  if (!isChainDotHead(dotOp))
+    return mlir::LLVM::AMD::isStoredContigWithMfmaLayout(
+        dotOp, axisAnalysisPass, mfmaEnc);
+
+  RankedTensorType oldRetType = dotOp.getType();
+  auto retShape = oldRetType.getShape();
+  unsigned rank = oldRetType.getRank();
+  // Set isTransposed to enable intra warp conversion for
+  // the mfma16x16 layout of a dot op, depending on whether
+  // its result is used by operand 0 or operand 1 of another dot op.
+  if (mfmaVersion == 4 && is16BitElemTy && mDim == 16 && nDim == 16 &&
+      rank == 2 && !hasPreShuffledScale) {
+    if (isChainDotHead(dotOp, 0u) &&
```

Comment:
You don't need to check isChainDotHead here since after the above return the only case left is chainDotHead.